File: LeetCode/257/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:

        def helper(node, acc, res) :
            if(node is None) :
                return res
            
            # Set the state
            acc.append(str(node.val))

            if(not node.left and not node.right) :
                res.append('->'.join(acc))
            else :
                # Go to next state
                helper(node.left, acc, res)
                helper(node.right, acc, res)

            # Reset the state
            acc.pop()

            return res
        
        ans = helper(root, [], [])
        return ans

File: LeetCode/257/main.cpp
#include <string>
#include <vector>

/**
 * Definition for a binary tree node
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
  void helper(TreeNode *node, string acc, vector<string> &res) {

    if (acc.size() > 0) {
      acc += ("->" + to_string(node->val));
    } else {
      acc += to_string(node->val);
    }

    if (!node->left && !node->right) {
      res.emplace_back(acc);
    }

    if (node->left) {
      helper(node->left, acc, res);
    }
    if (node->right) {
      helper(node->right, acc, res);
    }
  }

  vector<string> binaryTreePaths(TreeNode *root) {
    vector<string> res;
    helper(root, "", res);
    return res;
  }
};